src/isolation/fd.rs

```diff
@@ -1,33 +1,107 @@
 use std::{
-	collections::HashSet,
+	fmt,
 	os::fd::{FromRawFd, OwnedFd, RawFd},
 };
 
-#[derive(Default, Debug)]
+use slab::Slab;
+use yoke::{Yoke, erased::ErasedArcCart};
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct GuestFd(pub i32);
+
+impl fmt::Display for GuestFd {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		write!(f, "GuestFd({})", self.0)
+	}
+}
+
+impl From<usize> for GuestFd {
+	fn from(x: usize) -> Self {
+		// The following should never panic unless someone allocates more than i32::MAX file descriptors
+		Self(x.try_into().unwrap())
+	}
+}
+
+impl GuestFd {
+	fn is_standard(self) -> bool {
+		self.0 < 3
+	}
+
+	fn get(self) -> usize {
+		self.0.try_into().unwrap()
+	}
+}
+
+/// Description of what this guest file descriptor wraps:
+#[derive(Clone)]
+pub enum FdData {
```

Comment:
An alternative (but I don't know if also a better) approach: 
We could make add a `GuestFile` trait that is implemented by  `RawFd` and `VirtualFd` structs, and store these in the `UhyveFileDescriptorLayer`. Then we could move the `read`,  `write`, etc. functions into the respective structs instead of having a big function in `hypercall.rs`.